File: bexar_clean.py
"""
Bexar County Criminal Court Data — Merge, Clean, and Export
Run from the directory containing the CSV files.
Outputs: bexar_cleaned.parquet, bexar_panel_1990_2021.parquet, bexar_cleaned.csv
"""

import re
import glob
import pandas as pd
import numpy as np
from pathlib import Path

DATA_DIR = Path(".")  # adjust if CSVs are in a subdirectory


# ── 1. Load and merge ─────────────────────────────────────────────────────────

def load_csvs(data_dir: Path) -> pd.DataFrame:
    files = sorted(glob.glob(str(data_dir / "DC_cjjorad_*.csv")))
    if not files:
        raise FileNotFoundError(f"No DC_cjjorad_*.csv files found in {data_dir}")
    print(f"Found {len(files)} CSV files: {[Path(f).name for f in files]}")

    frames = []
    reference_cols = None
    for f in files:
        df = pd.read_csv(f, dtype=str, low_memory=False)
        df = df.drop(columns=[c for c in df.columns if c.startswith("Unnamed:")], errors="ignore")
        if reference_cols is None:
            reference_cols = list(df.columns)
        else:
            if list(df.columns) != reference_cols:
                extra = set(df.columns) - set(reference_cols)
                missing = set(reference_cols) - set(df.columns)
                print(f"  WARNING schema drift in {Path(f).name}: extra={extra}, missing={missing}")
        frames.append(df)
        print(f"  Loaded {Path(f).name}: {len(df):,} rows")

    merged = pd.concat(frames, ignore_index=True)
    print(f"\nMerged total: {len(merged):,} rows, {len(merged.columns)} columns")
    return merged


def check_duplicates(df: pd.DataFrame) -> pd.DataFrame:
    key = "CASE-CAUSE-NBR"
    dups = df[df.duplicated(key, keep=False)]
    if len(dups):
        print(f"WARNING: {len(dups):,} rows share duplicate {key}. Keeping first occurrence.")
        df = df.drop_duplicates(key, keep="first")
    else:
        print(f"No duplicate {key} values — clean primary key.")
    return df


# ── 2. Strip whitespace ───────────────────────────────────────────────────────

def strip_strings(df: pd.DataFrame) -> pd.DataFrame:
    str_cols = df.select_dtypes("object").columns
    df[str_cols] = df[str_cols].apply(lambda s: s.str.strip())
    return df


# ── 3. Parse dates ────────────────────────────────────────────────────────────

DATE_COLS = ["CASE-DATE", "OFFENSE-DATE", "DISPOSITION-DATE",
             "CUSTODY-DATE", "SENTENCE-START-DATE", "SENTENCE-END-DATE", "BIRTHDATE"]

def parse_dates(df: pd.DataFrame) -> pd.DataFrame:
    for col in DATE_COLS:
        if col in df.columns:
            df[col] = pd.to_datetime(df[col], errors="coerce", infer_datetime_format=True)
    df["CASE-YEAR"] = df["CASE-DATE"].dt.year
    df["CASE-DECADE"] = (df["CASE-YEAR"] // 10 * 10).astype("Int64")
    df["DAYS-TO-CASE"] = (df["CASE-DATE"] - df["OFFENSE-DATE"]).dt.days
    return df


# ── 4. Recode categorical variables ──────────────────────────────────────────

def recode_race(df: pd.DataFrame) -> pd.DataFrame:
    mapping = {"B": "Black", "W": "White", "L": "Latino"}
    df["RACE-LABEL"] = df["RACE"].map(mapping).fillna("Other/Unknown")
    return df


def recode_sex(df: pd.DataFrame) -> pd.DataFrame:
    df["SEX-LABEL"] = df["SEX"].map({"M": "M", "F": "F"}).fillna("Unknown")
    return df


def recode_attorney(df: pd.DataFrame) -> pd.DataFrame:
    col = "ATTORNEY-APPOINTED-RETAINED"
    if col not in df.columns:
        return df
    df["ATTORNEY-TYPE"] = df[col].map({
        "A": "Appointed",
        "H": "Hired",
        "S": "ProSe",
    }).fillna(pd.NA)  # blank → missing
    return df


def recode_offense_type(df: pd.DataFrame) -> pd.DataFrame:
    col = "OFFENSE-TYPE"
    if col not in df.columns:
        return df

    felony_map = {"F1": "F1", "F2": "F2", "F3": "F3", "FS": "FS"}
    misdemeanor = {"MA", "MB", "MC", "M "}

    def classify(v):
        if pd.isna(v) or v == "":
            return "Unknown"
        if v in felony_map:
            return felony_map[v]
        if v in misdemeanor:
            return "Misdemeanor"
        return "Other Felony"

    df["OFFENSE-CLASS"] = df[col].apply(classify)
    # ordinal severity for felonies (lower = more serious)
    severity = {"F1": 1, "F2": 2, "F3": 3, "FS": 4,
                "Other Felony": 5, "Misdemeanor": 6, "Unknown": np.nan}
    df["OFFENSE-SEVERITY"] = df["OFFENSE-CLASS"].map(severity)
    return df


# ── 5. Parse ORIGINAL-SENTENCE ────────────────────────────────────────────────

SENT_RE = re.compile(
    r"(\d+)\s*YR.*?(\d+)\s*MTH.*?(\d+)\s*DYS",
    re.IGNORECASE,
)

def parse_sentence(s: str) -> float | None:
    if pd.isna(s) or s.strip() == "":
        return np.nan
    m = SENT_RE.search(s)
    if not m:
        return np.nan
    yrs, mths, days = int(m.group(1)), int(m.group(2)), int(m.group(3))
    total = yrs * 365.25 + mths * 30.44 + days
    return total


def parse_sentences(df: pd.DataFrame) -> pd.DataFrame:
    col = "ORIGINAL-SENTENCE"
    if col not in df.columns:
        return df
    df["SENTENCE-DAYS"] = df[col].apply(parse_sentence)
    df["SENTENCE-ZERO"] = df["SENTENCE-DAYS"] == 0  # probation-only / deferred
    return df


# ── 6. Outcome indicator variables ───────────────────────────────────────────

def build_outcomes(df: pd.DataFrame) -> pd.DataFrame:
    disp = df.get("DISPOSITION-DESC", pd.Series("", index=df.index)).fillna("")
    judg = df.get("JUDGEMENT-DESC", pd.Series("", index=df.index)).fillna("")

    df["DEFERRED"]    = disp.str.contains("DEFR", case=False, na=False).astype(int)
    df["DISMISSED"]   = disp.str.contains("DSMD", case=False, na=False).astype(int)
    df["GUILTY-PLEA"] = disp.str.contains("CT-GUILTY", case=False, na=False).astype(int)
    df["JURY-TRIAL"]  = (
        disp.str.contains("PNG JRY", case=False, na=False) |
        judg.str.contains("JURY",    case=False, na=False)
    ).astype(int)
    df["PROBATION"]   = disp.str.contains("PROB", case=False, na=False).astype(int)
    df["CONVICTED"]   = (
        (df["GUILTY-PLEA"] == 1) |
        (df["JURY-TRIAL"]  == 1) |
        (~df["SENTENCE-DAYS"].isna() & (df["SENTENCE-DAYS"] > 0))
    ).astype(int)
    return df


# ── 7. Bond amount ────────────────────────────────────────────────────────────

def clean_bond(df: pd.DataFrame) -> pd.DataFrame:
    col = "BOND-AMOUNT"
    if col not in df.columns:
        return df
    df[col] = pd.to_numeric(df[col], errors="coerce")
    df["BOND-MISSING"] = (df[col].isna() | (df[col] == 0)).astype(int)
    df["BOND-OUTLIER"] = (df[col] > 1_000_000).astype(int)
    return df


# ── 8. Prosecutor names ───────────────────────────────────────────────────────

def clean_prosecutors(df: pd.DataFrame) -> pd.DataFrame:
    for col in ["INTAKE-PROSECUTOR", "OUTTAKE-PROSECUTOR"]:
        if col in df.columns:
            df[col] = df[col].str.upper().str.strip()
    # case counts per intake prosecutor
    if "INTAKE-PROSECUTOR" in df.columns:
        counts = df["INTAKE-PROSECUTOR"].value_counts()
        df["PROSECUTOR-CASE-COUNT"] = df["INTAKE-PROSECUTOR"].map(counts)
    return df


# ── 9. Main ───────────────────────────────────────────────────────────────────

def main():
    df = load_csvs(DATA_DIR)
    df = check_duplicates(df)
    df = strip_strings(df)
    df = parse_dates(df)
    df = recode_race(df)
    df = recode_sex(df)
    df = recode_attorney(df)
    df = recode_offense_type(df)
    df = parse_sentences(df)
    df = build_outcomes(df)
    df = clean_bond(df)
    df = clean_prosecutors(df)

    # Historical frame (pre-1990)
    df_historical = df[df["CASE-YEAR"] < 1990].copy()
    print(f"\nHistorical records (pre-1990): {len(df_historical):,}")

    # Main panel
    panel = df[(df["CASE-YEAR"] >= 1990) & (df["CASE-YEAR"] <= 2021)].copy()
    print(f"Main panel (1990–2021): {len(panel):,}")

    # Save full cleaned dataset
    df.to_parquet("bexar_cleaned.parquet", index=False)
    df.to_csv("bexar_cleaned.csv", index=False)
    print("\nSaved: bexar_cleaned.parquet, bexar_cleaned.csv")

    # Save main panel
    panel.to_parquet("bexar_panel_1990_2021.parquet", index=False)
    print("Saved: bexar_panel_1990_2021.parquet")

    # Quick sanity checks
    print("\n── Race distribution (panel) ──")
    print(panel["RACE-LABEL"].value_counts())
    print("\n── Attorney type (panel) ──")
    print(panel["ATTORNEY-TYPE"].value_counts(dropna=False))
    print("\n── Offense class (panel) ──")
    print(panel["OFFENSE-CLASS"].value_counts())
    print("\n── Outcome rates (panel, primary races) ──")
    primary = panel[panel["RACE-LABEL"].isin(["Black", "White", "Latino"])]
    for outcome in ["DEFERRED", "DISMISSED", "GUILTY-PLEA", "CONVICTED"]:
        rates = primary.groupby("RACE-LABEL")[outcome].mean().round(4)
        print(f"\n{outcome}:\n{rates}")


if __name__ == "__main__":
    main()

File: bexar_clean.R
# Bexar County Criminal Court Data — Merge, Clean, Export
# Run from the directory containing the CSV files.
# Outputs: bexar_cleaned.parquet, bexar_panel_1990_2021.parquet, bexar_cleaned.csv
#
# install.packages(c("tidyverse", "arrow", "lubridate"))

library(tidyverse)
library(arrow)
library(lubridate)

DATA_DIR  <- "C:/Users/carol/Box/Bigelow/Bexar/Data"
OUTPUT_DIR <- "C:/Users/carol/Box/Bigelow/Bexar/Data"  # where .parquet/.csv outputs go

dir.create(OUTPUT_DIR, showWarnings = FALSE, recursive = TRUE)

# ── 1. Load and merge ─────────────────────────────────────────────────────────

files <- list.files(DATA_DIR, pattern = "^DC_cjjorad_", full.names = TRUE)
if (length(files) == 0) stop("No DC_cjjorad_*.csv files found in ", DATA_DIR)
message("Found ", length(files), " CSV files: ", paste(basename(files), collapse = ", "))

frames <- map(files, function(f) {
  df <- read_csv(f, col_types = cols(.default = "c"), show_col_types = FALSE)
  df <- select(df, -starts_with("Unnamed"))
  message("  Loaded ", basename(f), ": ", nrow(df), " rows, ", ncol(df), " cols")
  df
})

# Check schema consistency
col_sets <- map(frames, colnames)
ref_cols  <- col_sets[[1]]
walk2(files, col_sets, function(f, cols) {
  extra   <- setdiff(cols, ref_cols)
  missing <- setdiff(ref_cols, cols)
  if (length(extra) | length(missing))
    message("  WARNING schema drift in ", basename(f),
            " extra=", paste(extra, collapse=","),
            " missing=", paste(missing, collapse=","))
})

df <- bind_rows(frames)
message("\nMerged total: ", nrow(df), " rows, ", ncol(df), " columns")

# ── 2. Duplicate check ────────────────────────────────────────────────────────

key <- "CASE-CAUSE-NBR"
n_dups <- sum(duplicated(df[[key]]))
if (n_dups > 0) {
  message("WARNING: ", n_dups, " duplicate ", key, " rows. Keeping first.")
  df <- df[!duplicated(df[[key]]), ]
} else {
  message("No duplicate ", key, " values — clean primary key.")
}

# ── 3. Strip whitespace from all character columns ───────────────────────────

df <- df |> mutate(across(where(is.character), str_trim))

# ── 4. Parse dates ────────────────────────────────────────────────────────────

date_cols <- c("CASE-DATE", "OFFENSE-DATE", "DISPOSITION-DATE",
               "CUSTODY-DATE", "SENTENCE-START-DATE", "SENTENCE-END-DATE", "BIRTHDATE")

for (col in intersect(date_cols, colnames(df))) {
  df[[col]] <- parse_date_time(df[[col]], orders = c("mdy", "ymd", "dmy"), quiet = TRUE)
}

df <- df |>
  mutate(
    `CASE-YEAR`   = year(`CASE-DATE`),
    `CASE-DECADE` = floor(`CASE-YEAR` / 10) * 10L,
    `DAYS-TO-CASE` = as.numeric(difftime(`CASE-DATE`, `OFFENSE-DATE`, units = "days"))
  )

# ── 5. Recode RACE ────────────────────────────────────────────────────────────

df <- df |>
  mutate(`RACE-LABEL` = case_match(
    RACE,
    "B" ~ "Black",
    "W" ~ "White",
    "L" ~ "Latino",
    .default = "Other/Unknown"
  ))

# ── 6. Recode SEX ─────────────────────────────────────────────────────────────

df <- df |>
  mutate(`SEX-LABEL` = case_match(
    SEX,
    "M" ~ "M",
    "F" ~ "F",
    .default = "Unknown"
  ))

# ── 7. Recode ATTORNEY-APPOINTED-RETAINED ────────────────────────────────────

df <- df |>
  mutate(`ATTORNEY-TYPE` = case_match(
    `ATTORNEY-APPOINTED-RETAINED`,
    "A" ~ "Appointed",
    "H" ~ "Hired",
    "S" ~ "ProSe",
    .default = NA_character_
  ))

# ── 8. Recode OFFENSE-TYPE ───────────────────────────────────────────────────

df <- df |>
  mutate(
    `OFFENSE-CLASS` = case_match(
      `OFFENSE-TYPE`,
      "F1" ~ "F1",
      "F2" ~ "F2",
      "F3" ~ "F3",
      "FS" ~ "FS",
      c("MA", "MB", "MC", "M ") ~ "Misdemeanor",
      .default = "Other Felony"
    ),
    `OFFENSE-SEVERITY` = case_match(
      `OFFENSE-CLASS`,
      "F1" ~ 1L, "F2" ~ 2L, "F3" ~ 3L, "FS" ~ 4L,
      "Other Felony" ~ 5L, "Misdemeanor" ~ 6L,
      .default = NA_integer_
    )
  )

# ── 9. Parse ORIGINAL-SENTENCE → total days ──────────────────────────────────

parse_sentence <- function(s) {
  # Format: NNNYRnnMTHnnDYSnnnHR  e.g. "002YR00MTH00DYS000HR"
  yr  <- as.numeric(str_extract(s, "(\\d+)(?=\\s*YR)",  group = 1))
  mth <- as.numeric(str_extract(s, "(\\d+)(?=\\s*MTH)", group = 1))
  dys <- as.numeric(str_extract(s, "(\\d+)(?=\\s*DYS)", group = 1))
  total <- replace_na(yr, 0) * 365.25 +
           replace_na(mth, 0) * 30.44 +
           replace_na(dys, 0)
  if_else(is.na(yr) & is.na(mth) & is.na(dys), NA_real_, total)
}

if ("ORIGINAL-SENTENCE" %in% colnames(df)) {
  df <- df |>
    mutate(
      `SENTENCE-DAYS` = parse_sentence(`ORIGINAL-SENTENCE`),
      `SENTENCE-ZERO` = `SENTENCE-DAYS` == 0 & !is.na(`SENTENCE-DAYS`)
    )
}

# ── 10. Outcome indicator variables ──────────────────────────────────────────

# Detect actual column names (case-insensitive, handle spaces vs dashes)
all_cols <- colnames(df)
disp_col <- all_cols[str_detect(str_to_upper(all_cols), "DISPOSITION.*DESC")][1]
judg_col <- all_cols[str_detect(str_to_upper(all_cols), "JUDG.*DESC")][1]

if (is.na(disp_col)) {
  message("WARNING: No DISPOSITION-DESC column found. DEFERRED/DISMISSED/GUILTY-PLEA will be NA.")
  df$`DISPOSITION-DESC` <- NA_character_
  disp_col <- "DISPOSITION-DESC"
}
if (is.na(judg_col)) {
  df$`JUDGEMENT-DESC` <- NA_character_
  judg_col <- "JUDGEMENT-DESC"
}

# Standardize to expected names
if (disp_col != "DISPOSITION-DESC") {
  df <- df |> rename(`DISPOSITION-DESC` = all_of(disp_col))
}
if (judg_col != "JUDGEMENT-DESC") {
  df <- df |> rename(`JUDGEMENT-DESC` = all_of(judg_col))
}

df <- df |>
  mutate(
    disp_ = replace_na(`DISPOSITION-DESC`, ""),
    judg_ = replace_na(`JUDGEMENT-DESC`,   ""),
    DEFERRED    = as.integer(str_detect(disp_, regex("DEFR",      ignore_case = TRUE))),
    DISMISSED   = as.integer(str_detect(disp_, regex("DSMD",      ignore_case = TRUE))),
    `GUILTY-PLEA` = as.integer(str_detect(disp_, regex("CT-GUILTY", ignore_case = TRUE))),
    `JURY-TRIAL`  = as.integer(
      str_detect(disp_, regex("PNG JRY", ignore_case = TRUE)) |
      str_detect(judg_, regex("JURY",    ignore_case = TRUE))
    ),
    PROBATION   = as.integer(str_detect(disp_, regex("PROB",      ignore_case = TRUE))),
    CONVICTED   = as.integer(
      `GUILTY-PLEA` == 1L | `JURY-TRIAL` == 1L |
      (!is.na(`SENTENCE-DAYS`) & `SENTENCE-DAYS` > 0)
    )
  ) |>
  select(-disp_, -judg_)

# ── 11. Bond amount ───────────────────────────────────────────────────────────

if ("BOND-AMOUNT" %in% colnames(df)) {
  df <- df |>
    mutate(
      `BOND-AMOUNT` = as.numeric(`BOND-AMOUNT`),
      # Sentinel value >= 9,999,990 means "no bond set / missing" — treat as NA
      `BOND-AMOUNT` = if_else(`BOND-AMOUNT` >= 9999990, NA_real_, `BOND-AMOUNT`),
      `BOND-MISSING` = as.integer(is.na(`BOND-AMOUNT`) | `BOND-AMOUNT` == 0),
      `BOND-OUTLIER` = as.integer(!is.na(`BOND-AMOUNT`) & `BOND-AMOUNT` > 1e6)
    )
}

# ── 12. Prosecutor names ──────────────────────────────────────────────────────

for (col in intersect(c("INTAKE-PROSECUTOR", "OUTTAKE-PROSECUTOR"), colnames(df))) {
  df[[col]] <- str_to_upper(str_trim(df[[col]]))
}

if ("INTAKE-PROSECUTOR" %in% colnames(df)) {
  prosecutor_counts <- df |>
    count(`INTAKE-PROSECUTOR`, name = "PROSECUTOR-CASE-COUNT")
  df <- left_join(df, prosecutor_counts, by = "INTAKE-PROSECUTOR")
}

# ── 13. Split and export ──────────────────────────────────────────────────────

df_historical <- df |> filter(`CASE-YEAR` < 1990)
message("Historical records (pre-1990): ", nrow(df_historical))

panel <- df |> filter(`CASE-YEAR` >= 1990, `CASE-YEAR` <= 2021)
message("Main panel (1990-2021): ", nrow(panel))

write_csv(df,    file.path(OUTPUT_DIR, "bexar_cleaned.csv"))
write_parquet(df,    file.path(OUTPUT_DIR, "bexar_cleaned.parquet"))
write_parquet(panel, file.path(OUTPUT_DIR, "bexar_panel_1990_2021.parquet"))
message("Saved: bexar_cleaned.csv, bexar_cleaned.parquet, bexar_panel_1990_2021.parquet")

# ── 14. Quick sanity checks ───────────────────────────────────────────────────

message("\n── Race distribution (panel) ──")
print(count(panel, `RACE-LABEL`, sort = TRUE))

message("\n── Attorney type (panel) ──")
print(count(panel, `ATTORNEY-TYPE`, sort = TRUE))

message("\n── Offense class (panel) ──")
print(count(panel, `OFFENSE-CLASS`, sort = TRUE))

message("\n── Outcome rates by race (primary races, panel) ──")
panel |>
  filter(`RACE-LABEL` %in% c("Black", "White", "Latino")) |>
  group_by(`RACE-LABEL`) |>
  summarise(
    N              = n(),
    Deferred_pct   = mean(DEFERRED,      na.rm = TRUE) * 100,
    Dismissed_pct  = mean(DISMISSED,     na.rm = TRUE) * 100,
    GuiltyPlea_pct = mean(`GUILTY-PLEA`, na.rm = TRUE) * 100,
    Convicted_pct  = mean(CONVICTED,     na.rm = TRUE) * 100,
    .groups = "drop"
  ) |>
  print()
